### PR TITLE
MAINT: stats: make reducing functions emit consistent warning when sample is too small or empty

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Address stats part of gh-19805

#### What does this implement/fix?
Eliminates inconsistent errors and excessive warnings emitted by most stats reducing functions when 1D input is too small. Documents required sample sizes. Emits a single warning when NaN outputs are generated without NaNs in the input or when `nan_policy='omit'`. 

Return values (usually NaNs) are already correct for the most part; this just deals with warnings and errors.

Examples below are for `skew`, but the behavior of most reducing functions is adjusted.

**Before**:
```python3
import numpy as np
from scipy import stats

stats.skew([])  # too noisy
# /usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:1193: RuntimeWarning: Mean of empty slice.
#   mean = a.mean(axis, keepdims=True)
# /usr/local/lib/python3.10/dist-packages/numpy/core/_methods.py:121: RuntimeWarning: invalid value encountered in divide
#   ret = um.true_divide(
# /usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
#   return _methods._mean(a, axis=axis, dtype=dtype,
# /usr/local/lib/python3.10/dist-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
#   ret = ret.dtype.type(ret / rcount)
# nan

stats.skew([np.nan], nan_policy='omit')  # too noisy
# /usr/local/lib/python3.10/dist-packages/scipy/stats/_stats_py.py:1193: RuntimeWarning: Mean of empty slice.
#   mean = a.mean(axis, keepdims=True)
# /usr/local/lib/python3.10/dist-packages/numpy/core/_methods.py:121: RuntimeWarning: invalid value encountered in divide
#   ret = um.true_divide(
# /usr/local/lib/python3.10/dist-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
#   return _methods._mean(a, axis=axis, dtype=dtype,
# /usr/local/lib/python3.10/dist-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
#   ret = ret.dtype.type(ret / rcount)
# nan

stats.skew([[]], axis=1)  # silent
# array([nan])

stats.skew([[np.nan]], nan_policy='omit', axis=1)  # silent
# array([nan])
```

**After**: (maybe I should change "is too small" → "contains too few observations")
```python3
import numpy as np
from scipy import stats

stats.skew([])
# UserWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.
#   stats.skew([])
# nan

stats.skew([np.nan], nan_policy='omit')
# After omitting NaNs, one or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.
#  stats.skew([np.nan], nan_policy='omit')
# nan

stats.skew([[]], axis=1)
# UserWarning: All axis-slices of one or more sample arguments are too small; all elements of returned arrays will be NaN. See documentation for sample size requirements.
#  stats.skew([[]], axis=1)
# array([nan])

stats.skew([[np.nan]], nan_policy='omit', axis=1)
# UserWarning: After omitting NaNs, one or more axis-slices of one or more sample arguments is too small; corresponding elements of returned arrays will be NaN. See documentation for sample size requirements.
#  stats.skew([[np.nan]], nan_policy='omit', axis=1)
# array([nan])
```

#### Additional information
Consider ignoring whitepace changes and reviewing the commits separately. Each has a distinct purpose summarized by the commit message.

@rgommers @h-vetinari all I'd ask for now (if you're still interested in gh-19805) is to 1) do some before/after testing to see whether you prefer the new behavior (almost all reducing functions are fair game) and 2) help me answer these two questions.

Questions before I continue:
- Initially, I've changed all the tests that looked for a function-specific error message or warning; now they look for the standard behavior. It took a lot of time, but I think it was important to investigate each case. That said, should I just eliminate these function-specific tests? I would want to add a generic test anyway, and that would make these scattered tests redundant.
- Should the generic test (which looks for the correct warning and return value) check *every function*, or should I just check, say, a few representative functions (e.g. `wilcoxon` with one argument, `mannwhitneyu` with two arguments, and `kruskal` with three arguments)?

To do:
- [x] fix special cases (e.g. `mode`, `f_oneway`)
- [x] fix `_axis_nan_policy` test - it's very broken because it actually looked for the warnings/errors produced by each function
- [x] add test that correct warning is emitted in each case
- [ ] add function-specific warning information

For a follow-up - fix for array-API.

We still need to adjust a lot of things about the documentation. Some documentation assumes the samples are 1D, whereas in others it is careful to refer to axis-slices of ND arrays. I would prefer that we write the API documentation using 1D language but consistently link to a tutorial that describes how `axis`/`nan_policy`/`keepdims` work with N-D arrays. That's for a separate PR.

Also, most hypothesis tests can easily get a `method` argument that computes a more accurate p-values instead of returning NaN for small samples (e.g. like [`pearsonr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.pearsonr.html) and [`anderson_ksamp`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.anderson_ksamp.html) now have). That can be a series of follow-up PRs.

### CODE REVIEW

**Comment by mdhaber** [comment]
@h-vetinari @rgommers If you could just suggest answers to these two questions for now and compare the before/after behavior, I'd appreciate it:

- Many stats functions had tests that looked for a function-specific error message or warning when there were too few observations. I changed each of these tests to look for the new, standard behavior instead. It took a lot of time, but I think it was important to investigate each case. That said, should I just eliminate these function-specific tests? I need to add a generic test that looks for the new behavior anyway, and that would make these scattered tests redundant.
- The generic test will looks for the correct warning and return value. Should it check _every function_, or should I just check, say, a few representative functions (e.g. `wilcoxon` with one argument, `mannwhitneyu` with two arguments, and `kruskal` with three arguments)?

File: scipy/stats/_morestats.py
Comment (h-vetinari) [inline scipy/stats/_morestats.py:3842]: Same here

File: scipy/stats/_stats_py.py
Comment (h-vetinari) [inline scipy/stats/_stats_py.py:1619]: Same here

File: scipy/stats/_stats_py.py
Comment (h-vetinari) [inline scipy/stats/_stats_py.py:1809]: Same here

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (h-vetinari) [inline scipy/stats/tests/test_axis_nan_policy.py:479]: These additions go in the wrong direction; or are these just cases we haven't investigated/fixed yet?

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (h-vetinari) [inline scipy/stats/tests/test_axis_nan_policy.py:892]: idem (same below)

**Review by h-vetinari — COMMENTED** [review]
Generally this looks really nice! :)

> I changed each of these tests to look for the new, standard behavior instead.

Love this!

> That said, should I just eliminate these function-specific tests? I need to add a generic test that looks for the new behavior anyway, and that would make these scattered tests redundant.

Fine by me to just do a generic test (modulo caveats below)

> Should it check _every function_, or should I just check, say, a few representative functions

Parametrization is easy and if the tests run fast, I don't see a problem to run this for (almost) every function.

File: scipy/stats/_hypotests.py
Comment (mdhaber) [reply scipy/stats/_hypotests.py:102]: We can probably keep them for for 1D input. What I'll do is try-except the function and append the error message to the warning or something. For ND I would prefer not to do this for now: no warning is emitted at all now, so emitting some sort of warning is only a step in the right direction.

Comment (mdhaber) [reply scipy/stats/_hypotests.py:102]: > but only adding it there and not enforcing it anymore is IMO a step back.

Wait - this makes me think that there might be a misunderstanding. These limitations are still enforced, but rather than raising a `ValueError` or return NaN and emit multiple warnings (which may or may not be informative), the function returns `nan` and emits a standard "small sample" warning that refers to the documentation for specific sample size limitations.

My first response https://github.com/scipy/scipy/pull/20694#discussion_r1607435544 still holds - for 1D input, we could append this function-specific information to the warning message without too much trouble. LMK if that's what you'd like to see.

File: scipy/stats/_morestats.py
Comment (mdhaber) [reply scipy/stats/_morestats.py:2006]: OK, and this makes me think that there is no misunderstanding : ) Yes, exactly - I wouldn't have to revamp the testing.

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/tests/test_axis_nan_policy.py:479]: I think these are fine because we emit these warnings intentionally, and looking for them is not the point of _this_ test, so it seems OK to ignore them.

Comment (h-vetinari) [reply scipy/stats/tests/test_axis_nan_policy.py:479]: Ok, if the warnings are intentional, then that's fine for me. I just want to avoid filtering out the _unintentional_ ones in the tests, rather doing it properly in the implementation

Comment (mdhaber) [reply scipy/stats/tests/test_axis_nan_policy.py:479]: I see. Yeah, these are the messages of the intentional warnings. Maybe I will import the variables that hold them to make it more clear.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:4998]: Here's what this comment is about:
```python3
import numpy as np
from scipy.stats import skew
skew(np.ones((0, 1)))  # array([nan]) and warning
# SmallSampleWarning: All axis-slices of one or more sample arguments are too small; all elements of returned arrays will be NaN. See documentation for sample size requirements.
skew(np.ones((0, 0)))  # array([], dtype=float64)
```
Both of these calls take the statistic along a slice with length 0. However, the result of the second one is an empty array, so no NaNs actually get produced, and the warning seems unnecessary, so I didn't think the output was surprising enough to deserve a warning. LMK if you disagree.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [inline scipy/stats/tests/test_stats.py:7728]: This is technically undesirable:
```python3
from scipy import stats
stats.f_oneway([])  # F_onewayResult(statistic=nan, pvalue=nan) and "too small" warning
stats.f_oneway([1, 2, 3])  # at least two inputs are required; got 1.
```
This probably affects all decorated functions that accepts `*args`. It happens because the decorator produces the warning and result before checking the number of arguments.

**Review by mdhaber — COMMENTED** [review]
Some comments to aid the reviewer.

**Comment by ilayn** [comment]
This one went a bit over my head to be honest. So I must leave this to you folks unfortunately. Apologies about that.

**Comment by mdhaber** [comment]
@ilayn If it helps, don't worry about the diff. If you're interested enough, just take a look at the example behavior in the top post. I don't think we need everyone to review line by line, but if you're willing to just play with a few stats functions with empty arguments before and after this PR to confirm that we like the new behavior better, that would be appreciated.

File: scipy/stats/tests/test_stats.py
Comment (h-vetinari) [reply scipy/stats/tests/test_stats.py:7728]: I guess we leave this for a follow-up? If so, could you open an issue about it?

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (h-vetinari) [reply scipy/stats/tests/test_axis_nan_policy.py:395]: This is still an open TODO, or am I misunderstanding your comment?

File: scipy/stats/_axis_nan_policy.py
Comment (h-vetinari) [inline scipy/stats/_axis_nan_policy.py:610]: I think this assignment is redundant? We only have one user in the rest of the function, which could just use `too_small_nd_omit` directly

**Review by h-vetinari — COMMENTED** [review]
I didn't go over it with a fine-toothed comb, but basically this LGTM. Just some minor comments.

**Comment by mdhaber** [comment]
Yes, those things are still open items. I should rework the test before we merge, and I might be able to fix the issue with too few *arguments* not emitting an error easily. I also said I would try to include the more detailed information in the error messages, but I'd like to save that for a follow-up to prevent this from continuing to accumulate merge conflicts. I'll try to finish by tomorrow.

File: scipy/stats/tests/test_stats.py
Comment (mdhaber) [reply scipy/stats/tests/test_stats.py:7728]: Fixed.

**Comment by mdhaber** [comment]
OK, I think this is now doing what I intended to do here. A few limitations:
- The issue with calling `f_oneway` with too few arguments is resolved, but the problem existed before this PR., and it applies to other functions that accepts `*args`, so I think there is a more general fix to be applied.
- It only emits the generic warning for now rather than appending a function-specific part of the message. I can try to append the function-specific part [as described above](https://github.com/scipy/scipy/pull/20694#discussion_r1607435544) in a follow-up.
- By design, this is going to produce some warnings where there were none before. Should we make the `SmallSampleWarning` public to make them easier to filter?
- It doesn't improve the behavior for other array API backends; those still emit warnings or raise errors as before. We'll want to rework the decorator significantly for array API support (and maybe actually modify most functions to handle `nan_policy` with n-D arrays/`axis` natively using something like gh-20363), and we'll fix this then.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:2001]:
```suggestion
    xp = array_namespace(a)

    s, _ = skewtest(a, axis, _no_deco=True)
```

**Comment by mdhaber** [comment]
This is ready, but I went ahead and re-milestoned. Don't want the new warnings to cause grief in the release process.

**Review by h-vetinari — APPROVED** [review]
I think this is a clear improvement and it would be a pity to miss the release (much less you having to keep fixing conflicts all the time). I'd vote for including this in 1.14

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (h-vetinari) [reply scipy/stats/tests/test_axis_nan_policy.py]: > This needs to be moved into the `else` block, among other things.

Can you comment about the status of this? And if follow-ups are necessary, open a new issue?

Comment (h-vetinari) [reply scipy/stats/tests/test_axis_nan_policy.py]: Not sure why this doesn't show up in the main thread.

**Review by h-vetinari — COMMENTED** [review]
Re-review just to make this thread more visible

File: scipy/stats/tests/test_axis_nan_policy.py
Comment (mdhaber) [reply scipy/stats/tests/test_axis_nan_policy.py]: Oh, weird. Probably doesn't show up because this refers to old code. At the time you commented originally, it still needed work, but it has since been rewritten. So no, no follow-ups here.

**Comment by mdhaber** [comment]
> Not sure why this doesn't show up in the main thread.

I clicked on the [link](https://github.com/scipy/scipy/pull/20694/files/bf686fc865e3669600b2af79b9fba6a20b0e4451#diff-14f4cfd9ee9321574bf892476e5554c891c95bf0552e57f0b4968d8bae506395) to the outdated commit and commented there. In short, that's all set.

I wouldn't oppose a squash merge if you'll help field complaints about the new warnings (although I think we're removing many more than we're adding, and the ones we're adding are informative).

**Comment by h-vetinari** [comment]
Bombs away!

**Comment by h-vetinari** [comment]
Could you please write a release-note, and make sure it shows up in https://github.com/scipy/scipy/pull/20784?

**Comment by mdhaber** [comment]
Sure; please commit https://github.com/scipy/scipy/pull/20784/files#r1616624810 if it looks good to you.

**Comment by h-vetinari** [comment]
> Sure; please commit https://github.com/scipy/scipy/pull/20784/files#r1616624810 if it looks good to you.

Thank you! I'll leave the committing to Tyler. :)

**Pull request merged**
